### PR TITLE
D8CORE-4521 Localist field widget for events urls

### PR DESCRIPTION
**Pull request by pookmish**
# READY FOR REVIEW

# Summary
- This is the completed field formatter for Localist event feeds.  

# Review By (Date)
- 11/19

# Criticality
- 10

# Urgency
- High

# Review Tasks

## Setup tasks and/or behavior to test

1. Testing instructions are here: https://github.com/SU-SWS/stanford_profile/pull/463

### Site Configuration Sync

- Is there a config:export in this PR that changes the config sync directory?

## Front End Validation
- [ ] Design is approved by @ user?
- [ ] HTML validation: Is the markup using the appropriate semantic tags and [passes validation](https://validator.w3.org/nu/)? Or, [QA request ticket created](https://github.com/SU-SWS/template_warehouse/blob/master/jira_templates/QA_request_template.txt)?
- [ ] Cross-browser testing: Has been performed? Or, [QA request ticket created](https://github.com/SU-SWS/template_warehouse/blob/master/jira_templates/QA_request_template.txt)?
- [ ] Automated accessibility: Scans performed? Or, [QA request ticket created](https://github.com/SU-SWS/template_warehouse/blob/master/jira_templates/QA_request_template.txt)?
- [ ] Manual accessibility: Manually tested? Or, [QA request ticket created](https://github.com/SU-SWS/template_warehouse/blob/master/jira_templates/QA_request_template.txt)?

### Code
- [ ] Are the naming conventions following our standards?
- [ ] Does the code have sufficient inline comments?
- [ ] Is there anything in this code that would be hidden or hard to discover through the UI?
- [ ] Are there any [code smells](https://blog.codinghorror.com/code-smells/)?
- [ ] Are tests provided? eg (unit, behat, or codeception)

### Code security
- [ ] Are all [forms properly sanitized](https://www.drupal.org/docs/8/security/drupal-8-sanitizing-output)?
- [ ] Any obvious [security flaws or new areas for attack](https://www.drupal.org/docs/8/security)?

## General
- [ ] Is there anything included in this PR that is not related to the problem it is trying to solve?
- [ ] Is the approach to the problem appropriate?

# Affected Projects or Products
- Does this PR impact any particular projects, products, or modules?

# Associated Issues and/or People
- JIRA ticket(s)
- Other PRs
- Any other contextual information that might be helpful (e.g., description of a bug that this PR fixes, new functionality that it adds, etc.)
- Anyone who should be notified? (`@mention` them here)

# Resources
- [AMP Tool](https://stanford.levelaccess.net/index.php)
- [Accessibility Manual Test Script](https://docs.google.com/document/d/1ZXJ9RIUNXsS674ow9j3qJ2g1OAkCjmqMXl0Gs8XHEPQ/edit?usp=sharing)
- [HTML Validator](https://validator.w3.org/)
- [Browserstack](https://live.browserstack.com/dashboard) and link to [Browserstack Credentials](https://asconfluence.stanford.edu/confluence/display/SWS/External+Account+Credentials)

### CODE REVIEW

File: src/Plugin/Field/FieldWidget/LocalistUrlWidget.php [hunk 1]
```diff
@@ -0,0 +1,444 @@
...
+
+      $value['filters']['days'] = '365';
+
+      // We may in the future have a configuration value
+      // to include the "distinct"key to our API call.
+      // This tries to find such a value,
+      // and applies the key if it finds it.
+      try {
+        $config = \Drupal::config('config_pages.type.stanford_events_importer');
+        if ($config->get('third_party_settings.localist_select_distinct')) {
```
Comment (pookmish) [inline src/Plugin/Field/FieldWidget/LocalistUrlWidget.php:234]: This should be on the field widget settings, not third party settings. The widget will then build the URL based on that setting, so the distinct gets saved in the database directly with the rest of the url. We don't want the user to be able to change the option. Put int into the `settingsForm` & `defaultSettings` methods. Also this module should never rely on any config pages.

File: src/Plugin/Field/FieldWidget/LocalistUrlWidget.php [hunk 2]
```diff
@@ -0,0 +1,444 @@
...
+      '#type' => 'select',
+      '#title' => $this->t('Content Must Match'),
+      '#default_value' => $query_parameters['match'] ?? NULL,
+      '#empty_option' => $this->t('At least one place or group, and one filter item'),
+      '#options' => [
+        'any' => $this->t('Any venue, group, or filter item'),
+        'all' => $this->t('At least one venue and group, and all filter items'),
+        'or' => $this->t('Any venue or group, and one filter item'),
+      ],
+
```
Comment (pookmish) [inline src/Plugin/Field/FieldWidget/LocalistUrlWidget.php:194]: remove empty space

**Review by pookmish — COMMENTED** [review]
The `distinct` need to be moved. doing that will require changes to the configs in the stanford_profile.

File: src/Plugin/Field/FieldWidget/LocalistUrlWidget.php (hunk 2 above)
Comment (imonroe) [reply src/Plugin/Field/FieldWidget/LocalistUrlWidget.php]: fixed

File: src/Plugin/Field/FieldWidget/LocalistUrlWidget.php (hunk 1 above)
Comment (imonroe) [reply src/Plugin/Field/FieldWidget/LocalistUrlWidget.php]: fixed.  I put a checkbox in the field settings form as well.

File: src/Plugin/Field/FieldWidget/LocalistUrlWidget.php [hunk 3]
```diff
@@ -0,0 +1,451 @@
...
+   */
+  public function settingsForm(array $form, FormStateInterface $form_state) {
+    $elements = parent::settingsForm($form, $form_state);
+    $elements['placeholder_url']['#access'] = FALSE;
+    $elements['placeholder_title']['#access'] = FALSE;
+    $elements['select_distinct'] = FALSE;
+    $elements['select_distinct'] = [
+      '#type' => 'checkbox',
+      '#title' => $this->t('Select Distinct'),
+      '#return_value' => TRUE,
```
Comment (pookmish) [inline src/Plugin/Field/FieldWidget/LocalistUrlWidget.php:89]: this shouldn't return true all the time.  You should have a `#default_value` that pull from the settings

File: src/Plugin/Field/FieldWidget/LocalistUrlWidget.php [hunk 4]
```diff
@@ -0,0 +1,451 @@
...
+        continue;
+      }
+
+      $value['filters']['days'] = '365';
+
+      // We may in the future have a configuration value
+      // to include the "distinct"key to our API call.
+      // This tries to find such a value,
+      // and applies the key if it finds it.
+      try {
```
Comment (pookmish) [inline src/Plugin/Field/FieldWidget/LocalistUrlWidget.php:240]: You don't need a `try` around this.

File: src/Plugin/Field/FieldWidget/LocalistUrlWidget.php (hunk 3 above)
Comment (imonroe) [reply src/Plugin/Field/FieldWidget/LocalistUrlWidget.php:89]: The default setting is set to FALSE on line 64 above.  According to the documentation, the `#return_value` key indicates what value to return when the checkbox is checked, which I set for the sake of clarity: 

image

Am I misunderstanding the documentation?

File: src/Plugin/Field/FieldWidget/LocalistUrlWidget.php (hunk 4 above)
Comment (imonroe) [reply src/Plugin/Field/FieldWidget/LocalistUrlWidget.php]: removed.

File: config/schema/stanford_fields.schema.yml [hunk 5]
```diff
@@ -17,3 +17,19 @@ field.widget.settings.datetime_year_only:
...
+      type: string
+      label: 'Base URL'
+    placeholder_url:
+      type: string
+      label: 'Placeholder URL'
+    placeholder_title:
+      type: string
+      label: 'Placeholder title'
+    select_distinct:
+      type: integer
```
Comment (pookmish) [inline config/schema/stanford_fields.schema.yml:34]:
```suggestion
      type: boolean
```

File: src/Plugin/Field/FieldWidget/LocalistUrlWidget.php (hunk 3 above)
Comment (pookmish) [reply src/Plugin/Field/FieldWidget/LocalistUrlWidget.php:89]: A checkbox will return `1` or `0` which then gets cast to a boolean with https://github.com/SU-SWS/stanford_fields/pull/21#discussion_r752518203 so you don't have to define what it returns
But there is still no `#default_value` here.

File: config/schema/stanford_fields.schema.yml (hunk 5 above)
Comment (imonroe) [reply config/schema/stanford_fields.schema.yml]: fixed.

File: src/Plugin/Field/FieldWidget/LocalistUrlWidget.php (hunk 3 above)
Comment (imonroe) [reply src/Plugin/Field/FieldWidget/LocalistUrlWidget.php]: Gotcha.  This is fixed now.

Comment (pookmish) [reply src/Plugin/Field/FieldWidget/LocalistUrlWidget.php]: Thanks.

**Pull request merged**